File: code-coaches/hard/security/Security.java
import java.util.Scanner;

public class Security {

    public static boolean isMoneySafe(String layout) {
        int moneyIndex = layout.indexOf('$');
        int thiefIndex = layout.indexOf('T');

        for (int i = 0; i < layout.length(); i++) {
            if (layout.charAt(i) == 'G' && (i < moneyIndex && i > thiefIndex || i > moneyIndex && i < thiefIndex)) {
                // Found a guard between the money and the thief
                return true;
            }
        }

        // No guard found between the money and the thief
        return false;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String layout = scanner.nextLine();

        if (isMoneySafe(layout)) {
            System.out.println("quiet");
        } else {
            System.out.println("ALARM");
        }

        scanner.close();
    }

}



// ALTERNATIVE VERSION
/*
import java.util.Scanner;

public class Security {

	public static void main(String[] args) {
		Scanner input = new Scanner(System.in);

		String diagram = input.nextLine();
		String[] diagramSplitted = diagram.split("\\$");

		boolean isThiefOnLeft = true;
		int thiefPos = -1;
		int guardPos = -1;

		int idx1 = 0;
		for (String part : diagramSplitted) {
			if (part.contains("T")) {
				int idx2 = 0;
				for (char c : part.toCharArray()) {
					if (Character.toUpperCase(c) == 'G') {
						if (!isThiefOnLeft && guardPos != -1) {
							continue;
						} else {
							guardPos = idx2;
						}
					} else if (Character.toUpperCase(c) == 'T') {
						thiefPos = idx2;
					}
				}
			} else if (idx1 == 0) {
				isThiefOnLeft = false;
				continue;
			}
		}

		boolean isLeftRobbed = (guardPos == -1 || guardPos < thiefPos) && isThiefOnLeft;
		boolean isRightRobbed = (guardPos == -1 || guardPos > thiefPos) && !isThiefOnLeft;

		if (isLeftRobbed || isRightRobbed) {
			System.out.print("ALARM");
		} else {
			System.out.print("quiet");
		}

		input.close();
	}

}
*/

File: code-coaches/hard/security/Security.swift
func isMoneySafe(_ layout: String) -> Bool {
    if let moneyIndex = layout.index(of: "$"),
        let thiefIndex = layout.index(of: "T") {
        for (i, char) in layout.enumerated() {
            let moneyDistance = layout.distance(from: moneyIndex, to: thiefIndex)
            let thiefDistance = layout.distance(from: thiefIndex, to: moneyIndex)
            let guardDistanceToMoney = layout.distance(from: moneyIndex, to: layout.index(layout.startIndex, offsetBy: i))
            let guardDistanceToThief = layout.distance(from: thiefIndex, to: layout.index(layout.startIndex, offsetBy: i))

            if char == "G" && ((i < moneyDistance && i > guardDistanceToThief) || (i > moneyDistance && i < guardDistanceToThief) || (i < thiefDistance && i > guardDistanceToMoney) || (i > thiefDistance && i < guardDistanceToMoney)) {
                // Found a guard between the money and the thief
                return true
            }
        }
    }

    // No guard found between the money and the thief
    return false
}

func main() {
    let layout = readLine()!

	if isMoneySafe(layout) {
		print("quiet")
	} else {
		print("ALARM")
	}
}

main()